demos/projects/PC/linux/tests/test_ca_recovery.c

```diff
@@ -0,0 +1,86 @@
+
+static char ucPayloadBuffer[ 2048 ];
+
+void prvCopyBuffer()
+{
+    memcpy( ucPayloadBuffer, ucTestPayload, strlen( ucTestPayload ) );
+}
+
+int vStartTestTask( void )
+{
+    AzureIoTResult_t xResult;
+    AzureIoTJSONReader_t xReader;
+    AzureIoTCARecovery_RecoveryPayload xPayload;
+
+    prvCopyBuffer();
+
+    AzureIoTJSONReader_Init( &xReader, ucPayloadBuffer, strlen( ucTestPayload ) );
+
+    printf( "Parsing Recovery Payload\n" );
+    xResult = AzureIoTCARecovery_ParseRecoveryPayload( &xReader, &xPayload );
+
+    if( xResult != eAzureIoTSuccess )
+    {
+        printf( "\tFailed!\n" );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+
+    printf( "Checking all struct values\n" );
+
+    if( strncmp( xPayload.pucIoTHubHostname, TEST_HUB_HOSTNAME, sizeof( TEST_HUB_HOSTNAME ) - 1 ) != 0 )
+    {
+        printf( "\tHub Name Failed!\n" );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+    else if( strncmp( xPayload.pucPayloadSignature, TEST_SIGNATURE, sizeof( TEST_SIGNATURE ) - 1 ) != 0 )
+    {
+        printf( "\tSignature Failed!\n" );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+    else if( strncmp( xPayload.pucTrustBundleJSONObjectText, TEST_BUNDLE_JSON_OBJECT_TEXT, sizeof( TEST_BUNDLE_JSON_OBJECT_TEXT ) - 1 ) != 0 )
+    {
+        printf( "\tTrust Bundle Object Text Failed!\n" );
+        printf( "\tReceived: %.*s", xPayload.ulTrustBundleJSONObjectTextLength, xPayload.pucTrustBundleJSONObjectText );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+    else if( strncmp( xPayload.xTrustBundle.pucVersion, TEST_BUNDLE_VERSION, sizeof( TEST_BUNDLE_VERSION ) - 1 ) != 0 )
+    {
+        printf( "\tBundle Version Failed!\n" );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+    else if( strncmp( xPayload.xTrustBundle.pucExpiryTime, TEST_BUNDLE_EXPIRY, sizeof( TEST_BUNDLE_EXPIRY ) - 1 ) != 0 )
+    {
+        printf( "\tBundle Expiry Failed!\n" );
+        return TEST_CA_RECOVERY_FAIL;
+    }
+    else if( strncmp( xPayload.xTrustBundle.pucCertificates, TEST_BUNDLE_CERT, sizeof( TEST_BUNDLE_CERT ) - 1 ) != 0 )
```

Comment:
The payload should be unescaped in-place, into the original binary form to allow the firmware flasher to use it as-is (no copies should be made since this could be quite large).

See https://github.com/Azure/azure-sdk-for-c/blob/feature/iot-certificate-management/sdk/samples/iot/paho_iot_provisioning_csr_trustbundle_sample.c#L484 for how this is handled in the TrustBundle feature.

Reply:
The JSON is unescaped in place in the middleware case. It isn't in the embedded SDKs, but this currently mirrors how ADU does it.